File: components/api-modules/src/main/resources/META-INF/dirigible/modules/src/platform/workspace.ts
/*
 * Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
import * as bytes from "@dirigible/io/bytes";
const WorkspaceFacade = Java.type("org.eclipse.dirigible.components.api.platform.WorkspaceFacade");

export function createWorkspace(name) {
	const native = WorkspaceFacade.createWorkspace(name);
	return new Workspace(native);
};

export function getWorkspace(name) {
	const native = WorkspaceFacade.getWorkspace(name);
	return new Workspace(native);
};

export function getWorkspacesNames() {
	const workspacesNames = WorkspaceFacade.getWorkspacesNames();
	if (workspacesNames) {
		return JSON.parse(workspacesNames);
	}
	return workspacesNames;
};

export function deleteWorkspace(name) {
	WorkspaceFacade.deleteWorkspace(name);
};

/**
 * Workspace object
 */
class Workspace {

	constructor(private native) { }

	getProjects() {
		const native = this.native.getProjects();
		return new Projects(native);
	};

	createProject(name) {
		const native = this.native.createProject(name);
		return new Project(native);
	};

	getProject(name) {
		const native = this.native.getProject(name);
		return new Project(native);
	};

	deleteProject(name) {
		this.native.deleteProject(name);
	};

	exists() {
		return this.native.exists();
	};

	existsFolder(path) {
		return this.native.existsFolder(path);
	};

	existsFile(path) {
		return this.native.existsFile(path);
	};

	copyProject(source, target) {
		this.native.copyProject(source, target);
	};

	moveProject(source, target) {
		this.native.moveProject(source, target);
	};

}

/**
 * Projects object
 */
class Projects {

	constructor(private native) { }

	size() {
		return this.native.size();
	};

	get(index) {
		const native = this.native.get(index);
		return new Project(native);
	};

}

/**
 * Project object
 */
class Project {

	constructor(private native) { }

	getName() {
		const collection = this.native.getInternal();
		return collection.getName();
	};

	getPath() {
		const collection = this.native.getInternal();
		return collection.getPath();
	};

	createFolder(path) {
		const native = this.native.createFolder(path);
		return new Folder(native);
	};

	exists() {
		return this.native.exists();
	};

	existsFolder(path) {
		return this.native.existsFolder(path);
	};

	getFolder(path) {
		const native = this.native.getFolder(path);
		return new Folder(native);
	};

	getFolders(path) {
		const native = this.native.getFolders(path);
		return new Folders(native);
	};

	deleteFolder(path) {
		return this.native.deleteFolder(path);
	};

	createFile(path, input) {
		const native = this.native.createFile(path, input);
		return new File(native);
	};

	existsFile(path) {
		return this.native.existsFile(path);
	};

	getFile(path) {
		const native = this.native.getFile(path);
		return new File(native);
	};

	getFiles(path) {
		const native = this.native.getFiles(path);
		return new Files(native);
	};

	deleteFile(path) {
		return this.native.deleteFile(path);
	};

}

/**
 * Folders object
 */
class Folders {

	constructor(private native) { }

	size() {
		const size = this.native.size();
		return size;
	};

	get(index) {
		const native = this.native.get(index);
		return new Folder(native);
	};

}

/**
 * Files object
 */
class Files {

	constructor(private native) { }

	size() {
		const size = this.native.size();
		return size;
	};

	get(index) {
		const native = this.native.get(index);
		return new File(native);
	};

}

/**
 * Folder object
 */
class Folder {

	constructor(private native) { }

	getName() {
		const collection = this.native.getInternal();
		const name = collection.getName();
		return name;
	};

	getPath() {
		const collection = this.native.getInternal();
		return collection.getPath();
	};

	createFolder(path) {
		const native = this.native.createFolder(path);
		return new Folder(native);
	};

	exists() {
		return this.native.exists();
	};

	existsFolder(path) {
		return this.native.existsFolder(path);
	};

	getFolder(path) {
		const native = this.native.getFolder(path);
		return new Folder(native);
	};

	getFolders(path) {
		const native = this.native.getFolders(path);
		return new Folders(native);
	};

	deleteFolder(path) {
		return this.native.deleteFolder(path);
	};

	createFile(path, input) {
		const native = this.native.createFile(path, input);
		return new File(native);
	};

	existsFile(path) {
		return this.native.existsFile(path);
	};

	getFile(path) {
		const native = this.native.getFile(path);
		return new File(native);
	};

	getFiles(path) {
		const native = this.native.getFiles(path);
		return new Files(native);
	};

	deleteFile(path) {
		return this.native.deleteFile(path);
	};

}

/**
 * File object
 */
class File {

	constructor(private native) { }

	getName() {
		const collection = this.native.getInternal();
		return collection.getName();
	};

	getPath() {
		const collection = this.native.getInternal();
		return collection.getPath();
	};

	getContentType() {
		return this.native.getContentType();
	};

	isBinary() {
		return this.native.isBinary();
	};

	getContent() {
		const output = WorkspaceFacade.getContent(this.native);
		if (output) {
			output;
		}
		return output;
	};

	getText() {
		const bytesOutput = this.getContent();
		return bytes.byteArrayToText(bytesOutput);
	};

	setContent(input) {
		const output = WorkspaceFacade.setContent(this.native, input);
		return output;
	};

	setText(input) {
		const bytesInput = bytes.textToByteArray(input);
		return this.setContent(bytesInput);
	};

	exists() {
		return this.native.exists();
	};
}

File: components/ide/ide-workspace/src/main/java/org/eclipse/dirigible/components/ide/workspace/domain/Workspace.java
/*
 * Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.dirigible.components.ide.workspace.domain;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.eclipse.dirigible.components.ide.workspace.json.WorkspaceGitHelper;
import org.eclipse.dirigible.repository.api.ICollection;
import org.eclipse.dirigible.repository.api.IRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * The Workspace's Workspace.
 */
public class Workspace extends Folder {
	
	/** The Constant logger. */
	private static final Logger logger = LoggerFactory.getLogger(Workspace.class);

	/**
	 * Instantiates a new workspace.
	 *
	 * @param workspaceCollection
	 *            the workspace collection
	 */
	public Workspace(ICollection workspaceCollection) {
		super(workspaceCollection);
	}

	/**
	 * Creates the project.
	 *
	 * @param name the name
	 * @return the i project
	 */
	public Project createProject(String name) {
		ICollection collection = this.createCollection(name);
		return new Project(collection);
	}

	/**
	 * Gets the project.
	 *
	 * @param name the name
	 * @return the project
	 */
	public Project getProject(String name) {
		ICollection collection = this.getCollection(name);
		return new Project(collection);
	}

	/**
	 * Gets the projects.
	 *
	 * @return the projects
	 */
	public List<Project> getProjects() {
		List<Project> projects = new ArrayList<Project>();
		List<ICollection> collections = this.getCollections();
		for (ICollection collection : collections) {
			projects.add(new Project(collection));
		}
		return projects;
	}

	/**
	 * Delete project.
	 *
	 * @param name the name
	 */
	public void deleteProject(String name) {
		ICollection projectCollection = this.getCollection(name);
		java.io.File gitFolder = WorkspaceGitHelper.getGitFolderForProject(projectCollection.getRepository(), projectCollection.getPath());
		if (gitFolder != null
				&& gitFolder.exists()) {
			try {
				org.apache.commons.io.FileUtils.deleteDirectory(gitFolder.getParentFile());
			} catch (Exception e) {
				if (logger.isErrorEnabled()) {logger.error(e.getMessage(), e);}
			}
		}
		this.removeCollection(name);
	}

	/**
	 * Copy project.
	 *
	 * @param sourceProject the source project
	 * @param targetProject the target project
	 */
	public void copyProject(String sourceProject, String targetProject) {
		ICollection collection = this.createCollection(sourceProject);
		collection.copyTo(this.getCollection(targetProject).getPath());
	}

	/**
	 * Copy folder.
	 *
	 * @param sourceProject the source project
	 * @param sourceFolderPath the source folder path
	 * @param targetProject the target project
	 * @param targetFolderPath the target folder path
	 */
	public void copyFolder(String sourceProject, String sourceFolderPath, String targetProject, String targetFolderPath) {
		ICollection collection = this.createCollection(sourceProject);
		collection.getCollection(sourceFolderPath).copyTo(this.getCollection(targetProject).getCollection(targetFolderPath).getPath());
	}

	/**
	 * Copy file.
	 *
	 * @param sourceProject the source project
	 * @param sourceFilePath the source file path
	 * @param targetProject the target project
	 * @param targetFilePath the target file path
	 */
	public void copyFile(String sourceProject, String sourceFilePath, String targetProject, String targetFilePath) {
		ICollection collection = this.createCollection(sourceProject);
		collection.getResource(sourceFilePath).copyTo(this.getCollection(targetProject).getResource(targetFilePath).getPath());
	}

	/**
	 * Move project.
	 *
	 * @param sourceProject the source project
	 * @param targetProject the target project
	 */
	public void moveProject(String sourceProject, String targetProject) {
		ICollection collection = this.createCollection(sourceProject);
		collection.moveTo(this.getCollection(targetProject).getPath());
	}

	/**
	 * Move folder.
	 *
	 * @param sourceProject the source project
	 * @param sourceFolderPath the source folder path
	 * @param targetProject the target project
	 * @param targetFolderPath the target folder path
	 */
	public void moveFolder(String sourceProject, String sourceFolderPath, String targetProject, String targetFolderPath) {
		ICollection collection = this.createCollection(sourceProject);
		collection.getCollection(sourceFolderPath).moveTo(this.getCollection(targetProject).getCollection(targetFolderPath).getPath());
	}

	/**
	 * Move file.
	 *
	 * @param sourceProject the source project
	 * @param sourceFilePath the source file path
	 * @param targetProject the target project
	 * @param targetFilePath the target file path
	 */
	public void moveFile(String sourceProject, String sourceFilePath, String targetProject, String targetFilePath) {
		ICollection collection = this.createCollection(sourceProject);
		collection.getResource(sourceFilePath).moveTo(this.getCollection(targetProject).getResource(targetFilePath).getPath());
	}

	/**
	 * Link project.
	 *
	 * @param sourceProject the source project
	 * @param targetPath the target path
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	public void linkProject(String sourceProject, String targetPath) throws IOException {
		this.getRepository().linkPath(this.getPath() + IRepository.SEPARATOR + sourceProject, targetPath);
		
	}

}
